### DISCUSSION TITLE
Using BigTreeTech SKR Pico

### DISCUSSION THREAD

**Discussion by shishu94** [grblHAL/RP2040#35] [Ideas]
I'm thinking about using this project with the SKR Pico board.
https://github.com/bigtreetech/SKR-Pico
It is a RP2040 based board initially intended for klipper in the 3D printing community. 
It has 4 stepper motor drivers, power regulation, PWM ports, and many other things and can be found relatively cheap.
Is there any attempt to build this driver for it ?

[Update] I noticed a mapping for it already so never mind the question.

**Comment by terjeio** [comment]
If you get around to to try/use this map please report back as I have not verified it myself.

**Reply by shishu94** [reply]
Sure, so far I was able to compile, upload and get the board recognized in IOSender. Limit sensors work. Was trying to use the PWM for the spindle couldn't make it work yet (I'm using a mini servo to debug, might connect a voltmeter or even the osciloscope if I manage to bring that to work lab). 
Also tried to connect it with lasergrbl and lightburn (the former being my main target) but that I couldn't find a way yet. The connection gets stuck. I'll need to understand what IOSender does that others don't, or what they expect that IOSender doesn't.
Didn't try the stepper motors yet. Need wire the main line of the board for that (was testing only with USB 5v).

I double checked the mapping of pins, they seem ok according to the pinout from SKR.

That's where I'm today. I'll try to do more tests during the weekend.

**Reply by shishu94** [reply]
Did a couple of tests more.

What is working:
- only works with IOSender
- X, Y, Z movement with UART trinamic control enabled. Super smooth movement. Tested 10k accel to 5000mm/min speed.
- Spindle control through PWM works perfectly.

What doesn't:
- Trinamic sensorless homing doesn't work. But that is normal, drivers are connected through UART not SPI.
- Still can't understand why laser grbl and lightburn can't connect. If that is something silly already known please share any insights ^^

Will keep testing and posting if I find anything new.

**Reply by terjeio** [reply]
>  But that is normal, drivers are connected through UART not SPI.

Tuning sensoless homing is tricky, more so with the UART versions of the TMC drivers. Have a look at the driver datasheet...

> Still can't understand why laser grbl and lightburn can't connect.

IIRC Lasergrbl assumes that it connects to a controller that has exactly the same settings as legacy Grbl or it crashes. It is nearly two years since I last patched it (v3.0.22) and I have no idea if this has changed in later versions. A quick test I did just now with the patched version resulted in it freezing with the RP2040 and working (at least it seems so) with ESP32 and STM32F756 controllers I happen to have on my desk. Odd.

Oh, I just tested with the latest version - it works with STM32F756 and hangs with RP2040...

LaserBurn is known to work with grblHAL so again odd. Is there something special with the RP2040 USB stack/driver? I do not have a LaserBurn license so I am not able to try.

You may try compiling with a different [compatibility level](https://github.com/grblHAL/core/wiki/Compatibility-level), but no sure this helps as Lasergrbl just freezes up for me (I have to terminate it from the task manager).

**Reply by shishu94** [reply]
I managed to test a few other things.

I checked that the Serial usb driver was not answering with a simple serial console.
So I recompiled with the uart serial Tx(0)/Rx(1) (disabling the USB serial) and plugged the board with a FTDI serial reader that I had laying around. Didn't mess with the compatibility level (it was at 0 by default anyway).

With that IOSender, LaserGRBL and Lightburn work out of the box. Could even run a laser cutting sequence with Lightburn for a more comprehensive test.
Will probably test on a generic Pico to check if I have the same behavior.

So I guess a problem is around the USB serial driver.  For today I won't be able to test more but will definitely dig around that in the next days.

**Reply by shishu94** [reply]
Turn out that the connection with Lightburn and probably LaserGRBL too is the DTR flag.
Just enabling it in Lightburn was enough to make everything work.

Currently the only remaining bug I have to solve is the M4 command which does behave weirdly.
In the end appart from the connection issues that are just configurations, everything works quitte well.

**Reply by terjeio** [reply]
> Currently the only remaining bug I have to solve is the M4 command which does behave weirdly.

It is possible a typo in the driver code can be the source of this, I have just commited a fix that may help. And thanks for the feedback, appreciated!

**Reply by brandonsherring** [reply]
Can I ask how you wired it up, I am having issues understanding the E0 spindle outputs and the cycle start and hold being on same pins.  Di you use the btt pico.map or did you compile it yourseld?  I am new to grblhal, been using klipper for years I decided to build a large format laser engraver for my first engraver..  Any assistance would be appreciated.
